include/seqan3/io/sam_file/detail/format_sam_base.hpp

```diff
@@ -433,182 +422,307 @@ inline void format_sam_base::read_header(stream_view_type && stream_view,
+                        }
+                        case make_tag('S', 'O'): // SO (sorting) tag
+                        {
+                            header_entry = std::addressof(hdr.sorting);
+                            break;
+                        }
+                        case make_tag('S', 'S'): // SS (sub-order) tag
+                        {
+                            header_entry = std::addressof(hdr.subsorting);
+                            break;
+                        }
+                        case make_tag('G', 'O'): // GO (grouping) tag
+                        {
+                            header_entry = std::addressof(hdr.grouping);
+                            break;
+                        }
+                        default: // unsupported header tag
+                        {
+                            print_cerr_of_unspported_tag("HD", raw_tag);
+                        }
                     }
 
-                    parse_tag_value(*who);
+                    if (header_entry != nullptr)
+                        parse_tag_value(*header_entry);
+                    else
+                        skip_until_predicate(is_char<'\t'> || is_char<'\n'>);
                 }
-                ++it;                                                   // skip newline
-            } break;
+                ++it; // skip newline
 
-            case 'S': /* SQ (sequence dictionary) tag */
+                if (hdr.format_version.empty())
+                    throw format_error{std::string{"The required VN tag in @HD is missing."}};
+
+                break;
+            }
+
+            case make_tag('S', 'Q'): // SQ (sequence dictionary) tag
             {
                 ref_info_present_in_header = true;
                 std::ranges::range_value_t<decltype(hdr.ref_ids())> id;
+                std::optional<int32_t> sequence_length{};
                 std::tuple<int32_t, std::string> info{};
 
-                parse_tag_value(id);                                    // parse required SN (sequence name) tag
-                ++it;                                                   // skip tab or newline
-                parse_tag_value(get<0>(info));                          // parse required LN (length) tag
-
-                if (is_char<'\t'>(*it))                                 // read rest of the tags
+                // All tags can appear in any order, SN and LN are required tags
+                while (is_char<'\t'>(*it))
                 {
-                    ++it;                                               // skip tab
-                    parse_until_newline(get<1>(info));
-                }
-                ++it;                                                   // skip newline
+                    ++it; // skip tab
+
+                    std::array<char, 2> raw_tag{};
+                    raw_tag[0] = *it;
+                    ++it;
+                    raw_tag[1] = *it;
+                    ++it;
```

Comment:
maybe we should put that into a function, as we do that multiple times:

```c++
auto read_tag = [](auto it, std::array<char, 2> & tag)
{
    raw_tag[0] = *it;
    ++it;
    raw_tag[1] = *it;
    ++it;
    return it;
}

std::array<char, 2> raw_tag{};
it = read_tag(it, raw_tag);
```

Reply:
why copy the iterator and not just capture ?

Reply:
We can also take it via lvalue. I'm always a bit confused, because the standard often returns an iterator as copy after an operation.

Reply:
I think I would prefer `[&it]` or `[&]` because we do it for other lambdas.
Maybe we could also have a function that directly returns the `uint16_t` tag since we need to do it anyways.

Reply:
So you mean to put these into the lambda function aswell?

```cpp
                    std::array<char, 2> raw_tag{};
                    it = read_tag(it, raw_tag);
                    uint16_t tag = make_tag(raw_tag[0], raw_tag[1]);
```

Reply:
so the idea is:
```cpp
std::array<char, 2> raw_tag{};

uint16_t get_tag = [*it](std::array<char, 2> & raw_tag)
{
    raw_tag[0] = *it;
    ++it;
    raw_tag[1] = *it;
    ++it;
    return make_tag(raw_tag[0], raw_tag[1]); 
}

switch (get_tag())
...
```

Reply:
Yeah basically, the array could also be once in the beginning, and then you could even do

```cpp
std::array<char, 2> raw_tag{}; // <- somwhere in the beginning and used every time when parsing a tag

// the lambda
uint16_t get_tag = [&] ()
{
    raw_tag[0] = *it;
    ++it;
    raw_tag[1] = *it;
    ++it;
    return make_tag(raw_tag[0], raw_tag[1]); 
}

// when parsing a tag
switch (get_tag())
```

but let's see what Marcel thinks

Reply:
I like that version, but I would still call it `parse_tag` to be in line with the other names.